File: Data Structures & Algorithms/reverse-a-linked-list/submission-3.py
import sys
from typing import Optional

sys.setrecursionlimit(6000)

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.rs
impl Solution {
    pub fn has_cycle(head: *mut ListNode) -> bool {
        if head.is_null() {
            return false;
        }

        let mut slow = head;
        let mut fast = head;

        loop {
            unsafe {
                if (*fast).next.is_null() || (*(*fast).next).next.is_null() {
                    return false;
                }
                
                slow = (*slow).next;
                fast = (*(*fast).next).next;

                if slow == fast {
                    return true;
                }
            }
        }
    }
}